src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs

```diff
@@ -567,5 +570,67 @@ protected virtual void GetObjectData(SerializationInfo info, StreamingContext co
         {
             throw new PlatformNotSupportedException();
         }
+
+        internal string DebuggerToString()
+        {
+            // DebuggerDisplayAttribute is inherited. Use virtual members instead of private fields to gather data.
+            int identitiesCount = 0;
+            foreach (var items in Identities)
+            {
+                identitiesCount++;
+            }
+
+            int claimsCount = 0;
+            foreach (var item in Claims)
+            {
+                claimsCount++;
+            }
+
+            // Return debug string optimized for the case of one identity.
+            if (identitiesCount == 1 && Identity is ClaimsIdentity claimsIdentity)
+            {
+                return $"Principal {claimsIdentity.DebuggerToString()}";
+            }
+
+            return $"Principal Identities Count: {identitiesCount}, Claims Count: {claimsCount}";
+        }
+
+        private sealed class ClaimsPrincipalDebugProxy
+        {
+            private readonly ClaimsPrincipal _principal;
+
+            public ClaimsPrincipalDebugProxy(ClaimsPrincipal principal)
+            {
+                _principal = principal;
+            }
+
+            public IEnumerable<Claim> Claims
+            {
+                get
+                {
+                    List<Claim> claims = new List<Claim>();
+                    foreach (var c in _principal.Claims)
+                    {
+                        claims.Add(c);
+                    }
+                    return claims;
+                }
+            }
```

Comment:
```suggestion
            public IEnumerable<Claim> Claims => new List<Claim>(_principal.Claims);
```
?

Reply:
I made the property `IEnumerable` because that's what the real `ClaimsPrincipal.Claims` property is.

I've never written a debug proxy before and the documentation about best practices is pretty light. Does it make a difference?

Reply:
> Does it make a difference?

Functionally, no, it's the actual type of the object that matters. But it might help a future reader to understand why you're doing `=> new List<Claim>(_principal.Claims)` instead of `=> _principal.Claims`.